worker/introspection/leases.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package introspection
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/juju/juju/core/lease"
+	"github.com/juju/juju/core/raftlease"
+	"gopkg.in/yaml.v2"
+)
+
+type leaseHandler struct {
+	leases Leases
+}
+
+// ServeHTTP is part of the http.Handler interface.
+func (h leaseHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ss, err := h.leases.Snapshot()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "error: %v\n", err)
+		return
+	}
+	snapshot, ok := ss.(*raftlease.Snapshot)
+	if !ok {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "expected *raftlease.Snapshot\n")
+		return
+	}
+	if snapshot.Version != 1 {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "only understand how to show version 1 snapshots\n")
+		return
+	}
+
+	data := h.translateSnapshot(snapshot)
+
+	q := r.URL.Query()
+	if v := q.Get("model"); v != "" {
+		data = h.filterModel(data, v)
+	}
+	if v := q["app"]; len(v) > 0 {
+		data = h.filterApp(data, v)
+	}
+
+	bytes, err := yaml.Marshal(h.format(data))
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "error: %v\n", err)
+		return
+	}
+	w.Write(bytes)
+}
+
+type leases struct {
+	// TODO: handle pinning
+	controller map[string]leaseInfo
+	models     map[string]map[string]leaseInfo
```

Comment:
perhaps a comment on what the map keys are

Reply:
added a comment.